lib/gruf/synchronized_client.rb

```diff
@@ -0,0 +1,102 @@
+#
+
+require 'concurrent'
+
+module Gruf
+  ##
+  # Ensures that we only have one active call to a given endpoint with a given set of params. This can be useful
+  # to mitigate thundering herds. Cache the result for a period of time so other requests, once unblocked, will get the
+  # same value. These results are expired after the given ttl.
+  #
+  class SynchronizedClient < Gruf::Client
+    ##
+    # Initialize the client and setup the stub
+    #
+    # @param [Module] service The namespace of the client Stub that is desired to load
+    # @param [Hash] options A hash of options for the client
+    # @option options [String] :password The password for basic authentication for the service.
+    # @option options [String] :hostname The hostname of the service. Defaults to linkerd.
+    # @option options [Float] :ttl How long to give until expiring results from previous calls (in seconds)
+    # @param [Hash] client_options A hash of options to pass to the gRPC client stub
+    #
+    def initialize(service:, options: {}, client_options: {})
+      @ttl = options[:ttl] || Gruf.client_cache_ttl
+      @ttl = @ttl.is_a?(Float) && @ttl > 0 ? @ttl : 5.0
+      @locks = Concurrent::Map.new
+      @cache = Concurrent::Map.new
+      super
+    end
+
+    ##
+    # Call the client's method with given params. If another call is already active for the same endpoint and the same
+    # params, block until the active call is complete. If a previously computed result is present in the cache, use it
+    # until it expires.
+    #
+    # @param [String|Symbol] request_method The method that is being requested on the service
+    # @param [Hash] params (Optional) A hash of parameters that will be inserted into the gRPC request message that is required
+    # for the given above call
+    # @param [Hash] metadata (Optional) A hash of metadata key/values that are transported with the client request
+    # @param [Hash] opts (Optional) A hash of options to send to the gRPC request_response method
+    # @return [Gruf::Response] The response from the server
+    # @raise [Gruf::Client::Error|GRPC::BadStatus] If an error occurs, an exception will be raised according to the
+    # error type that was returned
+    #
+    def call(request_method, params = {}, metadata = {}, opts = {}, &block)
+      # Generate a unique cache key based on the method and params
+      cache_key = "#{request_method}.#{params.hash}"
+
+      # Create a lock for this call if we haven't seen it already, then acquire it
+      lock = @locks.compute_if_absent(cache_key) { Mutex.new }
+      lock.synchronize do
+        # Get value from cache or create placeholder
+        response = @cache.get(cache_key)
+        if response
+          Gruf.logger.debug "Returning cached result for #{cache_key}"
+          next response
+        end
+
+        # Make the call and record it in the cache
+        response = super
```

Comment:
It might be worth it making a note as to what would happen should the client call here raise an exception

Reply:
I've already tested this, and `synchronize` will properly release the lock and the exception is raised to the caller.